File: src/@main/src/screen/levelup.js
const { send, src } = require("../..");
const { split, random } = require("../../common");
const { log } = require("../../commonServer");
const { args } = require("../chat/chat");

module.exports.predicate = ["!levelup", "!lvup", "!lvlup", "!level", "!lvl", "!lv"];
module.exports.permission = 0;
module.exports.execute = async (_reply, from, chatter, message, text, emote, reply) => {
    if (!src().screen.isScreenOn(_reply)) return [1, ""];
    let stat = args(text)?.[0]?.toLowerCase().trim();
    switch (stat) {
        case "hp":
        case "maxhp":
        case "con":
        case "health":
        case "tank":
        case "constitution":
            stat = "maxhp";
            break;
        case "damage":
        case "strength":
        case "str":
        case "dmg":
        case "atk":
        case "force":
        case "might":
        case "attack":
        case "aggression":
            stat = "attack";
            break;
        case "accuracy":
        case "acc":
        case "crit":
        case "critical":
        case "critchance":
        case "criticalchance":
            stat = "critchance";
            break;
        case "":
        case undefined:
            stat = random(["maxhp", "attack", "critchance"]);
            break;
        default:
            _reply("available stats are 'hp', 'attack' and 'crit'");
            return [0, ""];
    }
    let result = await send("gizmo", "levelup", chatter?.twitch?.name, stat);
    if (result[1]) {
        _reply(`Leveled up the guy's ${stat}`);
        return [0, stat];
    }
    else {
        _reply("you do not have a stat point, win a combat to gain stat points");
        return [1, ""];
    }
}

File: src/gizmo2/Gizmo/StreamOverlay/Commands/Shimeji/LevelUp.cs
namespace Gizmo.StreamOverlay.Commands.Shimeji
{
    public class LevelUp : Command
    {
        public override async Task<object?[]?> Execute(params object?[] args)
        {
            string name = WASD.Assert<string>(args[0]);
            string stat = WASD.Assert<string>(args[1]);
            if (name == null || stat == null) return [0, false];
            if (StreamOverlay.Shimeji.TryGetValue(name, out var i))
            {
                var sp = i.Get<int>("statpoint");
                if (sp <= 0) return [0, false];
                i.Set("statpoint", sp - 1);
                var value = stat switch
                {
                    "constitution" => 25,
                    "attack" => 2.5f,
                    "critchance" => 0.01f,
                    _ => 0
                };
                i.Set(stat, i.Get<float>(stat) + value);
                return [0, true];
            }
            return [0, false];
        }
    }
}
